### ISSUE TITLE
NodeController should only compare timestamps it generated locally

### ISSUE THREAD

**Issue by davidopp** [GoogleCloudPlatform/kubernetes#6193]
As some folks pointed out in the discussion for #5399, NodeController (in the mode where kubelet updates NodeStatus rather than NodeController probing kubelet health) is currently detecting node timeout based on comparing a timestamp written by kubelet (LastProbeTime in NodeReady condition in NodeStatus) vs. NodeController's own local clock. This doesn't work in the presence of clock skew. #6159 has been filed to address the clock skew issue more generally, but we can (and should) fix this specific problem in NodeController so that it is robust to clock skew.

The proposal is as follows: NodeController should maintain a map from node to timestamp, where timestamp is the last clock time on the NodeController when the NodeController saw the given node update its NodeStatus. For example, this expression (from MonitorNodeStatus())
```
nc.now().After(lastReadyCondition.LastProbeTime.Add(gracePeriod)) 
```
would use the value from the map rather than lastReadyCondition.LastProbeTime. (There are two more similar comparisons at the bottom of MonitorNodeStatus(), where we delete pods, that also need to change.) Of course we also have to add code to keep this map up-to-date as NodeController sees changes to NodeStatus.

Whenever a node is created (including when an existing node is re-added if master crashes and then comes back up), it is added to the map with a timestamp of whenever the node was added. Thus after startup, nodes will not be transitioned to NodeReady==Unknown until the timeout has passed. It's safe to be a bit aggressive in transitioning nodes to Unknown, because the timeout for actually evicting pods due to NodeReady==Unknown is much longer than the timeout for marking a node as NodeRead==Unknown (5m vs. 8s or 30s).

**Comment by davidopp** [comment]
cc @lavalamp @bgrant0607 @dchen1107 @erictune , who participated in formulating this proposal.

**Comment by davidopp** [comment]
Oh and just to be crystal clear, this map would be in-memory only, not written to etcd.

And we discussed some alternatives to this proposal, but they were not robust to running API server on a different machines from NodeController.

**Comment by davidopp** [comment]
> When there is no value in the map for a particular node, the value is interpreted as "now." 

Sorry, that should say "When there is no value in the map for a particular node, the value is interpreted as the time when NodeController was created." (I've edited the original comment.)

**Comment by ddysher** [comment]
How do you maintain the map, via another loop or in the same monitor code?   The problem can also be solved if we can expedite a watch based monitor.

**Comment by davidopp** [comment]
Currently NodeController uses polling; every 5 seconds it starts a goroutine (even if the last one didn't finish yet, unless I'm misunderstanding) that checks the node's NodeReady's LastProbeTime and changes NodeReady to ConditionUnknown if the LastProbeTime is too old. Since we're looking at each node in this loop, I don't think it would be any more inefficient to update the map entry for each node in this loop. We should fix things so that we don't run more than one iteration of this loop in parallel, though.

I don't think watch will work because we're looking for timeouts -- that is, we want to take an action when we *don't* hear from a node. If we only look at a node when we get a watch for its NodeStatus, then we will never detect a timeout. I think we have to do a polling loop like we have already.

**Comment by gmarek** [comment]
I think we can have background watch which updates aforementioned map and still loop through all nodes every 5 seconds.

I was considering something similar for handling retries in node controller, but thought that it's not necessary given that "hard" retries should be efficient enough.

**Comment by gmarek** [comment]
I'm slightly worried about another 'feature' we have now (and we'll have even after this change). Namely when machine on which NodeController runs will become unresponsive for some time (say podEvictionTimeout seconds) and then come back NodeController will evict everything everywhere.

Currently there's a slight chance that some part of kubelets will manage to update their statuses before they get killed (probably very very small part). After the change describe here there will be no chance at all, not that it makes any difference.

**Comment by gmarek** [comment]
To propose one solution: replicated master/node controller with some consensus algorithm protects from such scenario. The problem is that I think it'll be extremely hard to move to a replicated setup before v1.0

**Comment by gmarek** [comment]
@davidopp @thockin @brendanburns @bgrant0607 @dchen1107 @erictune

**Comment by erictune** [comment]
The node controller should ask the question *what is more likely: that the nodes I see as bad are really bad, or that I am bad*.

To keep it simple, assume node badness is equally and independently likely for all nodes. This can be refined later.

Probability that a specific node (the one the controller is on) is bad, at a moment in time is `p`
Probability that any `k` nodes are bad, out of a cluster of `n`nodes, is `b = BinomialProbability(n, k, p)`.

Node controller knows `n` and `k`.
When `n` and `k` are such that `p <= b`, then the other nodes are bad, so kill pods.   Otherwise, do not kill pods.

The trick is deciding what to set `p` to and properly approximating `BinomialProbability` for large `n`.

**Comment by gmarek** [comment]
As a 'workaround' we can add a logic which checks if NodeController got update from any Node, and if not assume that it's a problem with Controller itself. It's a bad idea for many different reasons.

**Comment by gmarek** [comment]
Eric was first, so I'll elaborate why I think it's a bad idea:
- complicates logic, needs tuning constants
- makes it impossible to distribute node processing into independent goroutines
- is not needed if we plan to add replication at some point

**Comment by erictune** [comment]
@gmarek taking your comments one at a time:
  * complicated logic is better than killing the wrong nodes.
  * Is there any evidence that the processing needs to be parallelized?
  *  The approach I described is still applicable if we add replica node controllers.  You first check if all the peers agree on which nodes are bad.  If they do, then you compare the probability that all replicas of node controller are bad with probability that `n` of `k` nodes are bad.

**Comment by gmarek** [comment]
@erictune
Yes complicated logic is better than killing wrong nodes, but I strongly believe that it should be avoided if possible. In this case we know that replication solves this problem completely (it's VERY unlikely that say, 3 or 5 "master" (API server, controllers, etcd) replicas will go down in the same time. In addition I heard that replication is on our plate (probably post v1.0 but not very far), thus I think we should very carefully consider not fixing this problem before adding replication. Yes it is bad, but I'm not sure it's worse than the alternative (i.e. adding some code very little people will understand).

**Comment by erictune** [comment]
Totally fine with not fixing this problem yet.

**Comment by timothysc** [comment]
Perhaps I'm missing something, but if we are serious about checking time stamps, why are we not enforcing 'stratum 2' clock requirements and be done with it?  

Any 'skew' would be negligible at the granularity of the afore mentioned checks?

**Comment by davidopp** [comment]
@timothysc Our thought here was that if there is a simple way to deal robustly with clock skew in NodeController, we should do it, in addition to something like what you described. And I think the proposal we have here actually *is* simple -- the only wrinkle is the one @gmarek identified ("when machine on which NodeController runs will become unresponsive for some time and then come back NodeController will evict everything everywhere"), but we have that same problem whether all the clocks in the system are synchronized or not.

@erictune I'm sure I'm being dense, but I didn't understand your proposal. It seems like something simple we could do that might be related to your proposal (I'm not suggesting to do it) is to put a watch on NodeStatus update in etcd, and have NodeController always assume that the maximum amount of time it might have been disconnected is now-X where X is the most recent time (relative to NodeController) when NodeController received a NodeStatus update from etcd. Then it would just add this value to the grace period and the pod eviction timeout, to be on the safe side. I think this might have been what @gmarek was proposing. Some bad things might happen if etcd gets backlogged, but I think that would also affect the loop where it's reading status and they would cancel out (NodeController might take longer than it should to detect that it became disconnected, but stuff would seem fresher than it really is).

BTW the way we handle this problem in our internal cluster management system is to rate limit deeming nodes to be gone/bad/whatever. The equivalent here would be to either change the logic for moving NodeReady to Unknown such that the required timeout increases as the number of nodes that have recently been moved to Unknown increases, or equivalently increase the pod eviction timeout due to Unknown state as the number of nodes that have recently been moved to Unknown increases. 

I don't think replicating the NodeController actually helps, because the problem here isn't a crash (we already said that when it comes back up, it makes the conservative assumption that it just heard from every node). The problem is some kind of temporary network disconnection or livelock, such that when it comes back, it doesn't reset anything in the map.

In any event, I don't think we should do anything fancy here. I would actually suggest doing nothing for now, and later adding rate limiting as I described. However, if you really want to do something in the mean time, I would just track the most recent NodeController time when NodeController read any status, and add now-that to podEvictionTimeout and to gracePeriod.

**Comment by ddysher** [comment]
@davidopp Only one goroutine is running in MonitorNodeStatus.  It just checks periodically, within the period, it is a sleep.  What I have in mind for watch solution is similar to this:
```go
for {
   switch {
   case <- watch.ResultChan()
       looks good to me, reset tick
   case <- tick(gracePeriod)
        bang!
   }
}
```

@gmarek I'm not entirely understand your concern.  When machine goes down (where NodeController is running) for podEvictionTimeout, and then comes back.  It will first ask apiserver for NodeStatus, and rebuild its knowledge of timestamp.  NodeController is basically stateless, I'm not sure why pod will be evicted in this case?  However, the problem will come out if we have an in-memory map, but similar to other places, we need to warm up the map at NodeController startup.  This also applies to the case where all master components crashes, as long as etcd data doesn't get lost, status can be reconstructed. If not, then we need to fix that.   I think a hot standby master should be equivalent to a restarted master.

**Comment by davidopp** [comment]
> Sorry, that should say "When there is no value in the map for a particular node, the value is interpreted
> as the time when NodeController was created." (I've edited the original comment.)

Sorry, I still got this wrong when I corrected it. It should say "Whenever a node is created, it is added to the map with a timestamp of whenever the machine was created."

**Comment by davidopp** [comment]
@ddysher He's not concerned about NodeController crashing and coming back up. He's concerned about the case where master becomes disconnected from the network and then reconnects say 10 minutes later. It will have queued pod deletes for every pod in the system during the last 5 minutes of those 10 minutes, which it will immediately write to etcd when it reconnects. As I mentioned, we fix this in our internal system using rate limiting.

I agree about warming up the map, and my two corrections in this thread were about how to do that properly. I think I finally got it right, and I updated the original description (first entry in the issue).

**Comment by gmarek** [comment]
We're discussing two slightly related bud different things here. Let's move "kill everything" discussion to #6228.

**Comment by ddysher** [comment]
@davidopp  oh, ok. master unresponsive and master go down is used in interchangablely, got confused.

**Comment by ddysher** [comment]
hmm, I was wondering where is ur comment.  Turns out we comment at the same time and I had to do a refresh to see your comment...

**Comment by davidopp** [comment]
To summarize
* I think you should add the map, so NodeController only uses timestamps it generated locally. Note the change in how the map entries get initialized, which I got wrong twice (I've updated the description in the first entry).
* I don't think we should try to fix the "kill everything" problem right now. We should add rate limiting later (even if "later" means "later this week" -- but let's do this one thing at a time.)

**Comment by gmarek** [comment]
Closed per #6217
